contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,212 @@
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called the ResourceType.
+It is defined as follows:
+
+```golang
+type ResourceType struct {
+	Kind string
+	Name string
+	Quantity resource.Quantity
+	Properties []string
+}
+```
+
+Because the current API (Capacity, Allocatable) can not be extended to
+support ResourceType, we will need to create two new attributes in the NodeStatus
+structure: `CapacityV2` and `AllocatabeV2`:
+
+```golang
+type NodeStatus struct {
+	Capacity ResourceList
+	CapacityV2 []ResourceType
+
+	Allocatable ResourceList
+	AllocatableV2 []ResourceType
+}
+```
+#### ResourceClass
+
+We will not allow the end user to specify `ResourceTypes` in his pod spec. Instead
+we propose a spec called `ResourceClass` which encapsulate devices requests.
+
+Only cluster admins will be able to define a ResourceClass by default. This can
+then be tweaked through RBAC.
+
+A ResourceClass spec will have the following spec:
+```yaml
+kind: ResourceClass
+apiVersion: compute.k8s.io/v1alpha1
+metadata:
+  name: nvidia-tesla-3.7
+spec:
+    resourceSelector:  // required are ORed
+        - required:  // all properties under required are ANDed
+          - key: kind
+            operator: In
+            values: ["nvidia-gpu"]
+          - key: memory
+            operator: Gt
+            values: ["8 Gi"]
+        - required:
+          - key: kind
+            operator: In
+            values: ["nvidia-gpu"]
+          - key: compute-capabilities
+            operator: Gt
+            values: ["3.7"]
+
+```
+
+The user will then be able to specify in his pod Spec requirements a
+Resource class through either Requests or Limits
+```yaml
+kind: Pod
+apiVersion: v1
+metadata:
+  name: gpu-pod
+spec:
+  containers:
+    name: gpu-container
+    resources:
+      limits:
+        cpu: 1.2
+        memory: 10Gi
+        nvidia-tesla-3.7: 4
+      requests:
+        cpu: 0.3
+```
+
+
+### Device Plugin
+
+Kubelet will create a gRPC connection (as a client) to the Device Plugins whose endpoint
+were specified on the CLI.
+
+The kubelet will call the `Discover` function once and then listen to a stream returned
+by the `Monitor` function.
+
+When receiving a pod which requests GPUs (through the ResourceClass api) kubelet will
+be in charge of deciding which device to assign and advertising the corresponding changes
+to the node's allocatable pool and pod assigned devices.
+The scheduler will still be in charge of filtering the nodes which cannot satisfy the
+requests.
+
+The typical process will follow the following pattern:
+1. A user submits a pod spec requesting X devices with Y properties
+2. The scheduler filters the nodes which do not match the requests
+3. The pod lands on the node
+4. Kubelet decides which device should be assigned to the pod
+5. Kubelet removes the devices from the `Allocatable` pool
+6. Kubelet updates the pod's status with the allocated devices
```

Comment:
User can know what devices were assigned to the pod. It can also be useful in the case of monitoring :)